<a href="https://colab.research.google.com/github/Srestrero/AlgorithmsUN2024II/blob/main/Lab3/srestreporo_Lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [7]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed bhi
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
ad

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [8]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [9]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

12
34
46

### Test input output files

In [10]:
%%writefile input1.txt
4 8

Writing input1.txt


In [11]:
%%writefile output1.txt
12

Writing output1.txt


In [12]:
%%writefile input2.txt
0 9

Writing input2.txt


In [13]:
%%writefile output2.txt
9

Writing output2.txt


In [14]:
!./APlusB < input1.txt

12

In [15]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [16]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [17]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [18]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [19]:
!python sum.py

12 34
46


In [20]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [21]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [22]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [23]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

^C


### Test 1: Small numbers

In [24]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [25]:
%%writefile output0.txt
0

Writing output0.txt


In [26]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [27]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [28]:
!./max_pairwise_product_full < input1.txt

90


In [29]:
%%writefile output1.txt
90

Overwriting output1.txt


In [30]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [31]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [32]:
%%writefile output2.txt
90

Overwriting output2.txt


In [33]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [34]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [35]:
200000*200000

40000000000

In [36]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [37]:
!./max_pairwise_product_full < input3.txt

1345294336


In [38]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [39]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [40]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [41]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [42]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [43]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [44]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [45]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [46]:
%%writefile input4.txt
12
1 700 4567892 32 12 4 90 3 200000 9 200000 23

Writing input4.txt


In [47]:
4567892 * 200000

913578400000

In [48]:
%%writefile output4.txt
913578400000

Writing output4.txt


In [49]:
%%writefile input5.txt
15
1 800 5678901 45 13 5 100 4 300000 10 300000 24 6000000 7000000 8

Writing input5.txt


In [50]:
7000000 * 6000000

42000000000000

In [51]:
%%writefile output5.txt
42000000000000

Writing output5.txt


In [52]:
%%writefile input6.txt
18
1 1000 8901234 50 15 6 150 5 500000 12 500000 30 9000000 10000000 9 12000000 15000000 7

Writing input6.txt


In [53]:
15000000 * 12000000

180000000000000

In [54]:
%%writefile output6.txt
180000000000000

Writing output6.txt


In [55]:
%%writefile input7.txt
20
2 2000 12345678 60 18 8 200 6 800000 15 800000 35 20000000 25000000 10 30000000 40000000 9 45000000 50000000

Writing input7.txt


In [56]:
50000000 * 45000000

2250000000000000

In [57]:
%%writefile output7.txt
2250000000000000

Writing output7.txt


In [58]:
%%writefile input8.txt
22
3 3000 23456789 70 20 10 300 7 1000000 18 1000000 40 50000000 60000000 12 70000000 80000000 11 90000000 100000000 110000000 120000000

Writing input8.txt


In [59]:
120000000 * 110000000

13200000000000000

In [60]:
%%writefile output8.txt
13200000000000000

Writing output8.txt


In [61]:
%%writefile input9.txt
25
4 4000 34567890 80 25 12 400 8 2000000 20 2000000 45 150000000 160000000 15 170000000 180000000 13 190000000 200000000 210000000 220000000 230000000 240000000 250000000

Writing input9.txt


In [62]:
250000000 * 240000000

60000000000000000

In [63]:
%%writefile output9.txt
60000000000000000

Writing output9.txt


In [64]:
%%writefile input10.txt
30
5 5000 45678901 100 30 15 500 10 3000000 25 3000000 50 300000000 350000000 18 400000000 450000000 16 500000000 600000000 650000000 700000000 750000000 800000000 850000000 900000000 950000000 999999999 1000000000 1000000000

Writing input10.txt


In [65]:
1000000000 * 1000000000

1000000000000000000

In [66]:
%%writefile output10.txt
1000000000000000000

Writing output10.txt


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [67]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [68]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [69]:
199999 * 200000

39999800000

In [70]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [71]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [72]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [73]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [74]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

^C


In [75]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

^C


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [76]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code




In [77]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [78]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [79]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [80]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [81]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [82]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [83]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [84]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [85]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

^C


In [86]:
# Estimated time by instruction

295/200003177092


1.4749765693187072e-09

In [87]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [88]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [89]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [90]:
10999 * 11000

120989000

In [91]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.859s
user	0m0.852s
sys	0m0.002s


## Python

In [92]:
%%writefile max_pairwise_product_execints.py
def MaxPairwiseProduct(numbers, execints):                      # INSTRUCTIONS executed by line
    max_product = 0                                             # 1
    execints[0] += 1

    aux_product = 0                                             # 1
    execints[0] += 1

    n = len(numbers)                                            # 1
    execints[0] += 1

    execints[0] += 1
    for first in range(n):                                      # 2 + 2n
        execints[0] += 2

        execints[0] += 1
        for second in range(n):                                 # 2n + 2n**2
            execints[0] += 2

            execints[0] += 1
            if first != second:                                 # n**2
                aux_product = numbers[first] * numbers[second]   # n**2 - n
                execints[0] += 1

                execints[0] += 1
                if aux_product > max_product:                    # n**2 - n
                    max_product = aux_product                    # [0, n**2/2 - n/2] upper bound
                    execints[0] += 1

        execints[0] += 1
    execints[0] += 1

    return max_product                                          # 5 + 2n + [5n**2, 11n**2/2 - n/2] upper bound

def main():                                                     # INSTRUCTIONS executed by line
    execints = [0]                                             # Usando lista para simular paso por referencia

    n = int(input())                                           # 1
    execints[0] += 1

    numbers = list(map(int, input().split()))                  # n
    execints[0] += n

    execints[0] += 1
    result = MaxPairwiseProduct(numbers, execints)             # 1 + 5 + 2n + [5n**2, 11n**2/2 - n/2]

    execints[0] += 1
    print(execints[0])                                         # Imprime el número total de instrucciones ejecutadas

    return 0                                                   # --------------
                                                              # 11 + 6n + [5n**2, 11n**2/2 - n/2] upper bound

if __name__ == '__main__':
    main()

Writing max_pairwise_product_execints.py


La complejidad temporal sigue siendo la misma:
$$T(n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2} - \frac{n}{2}\right) > 6n^2 = O(n^2)$$

In [93]:
# Verify
!python3 max_pairwise_product_execints.py < input0.txt

538


In [94]:
!python3 max_pairwise_product_execints.py < input2.txt

539


# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [95]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [96]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


The number of instructions executed by the algorithm with a list of size $n$ is  

$$T (n) = 11 + 7n + \left[2(n-1)n+1, 2.5(n-1)n\right)= O(n^2)$$

Then for the input0.txt
$$T = 11 + 7(10) + 2(9)11   = 279 $$

### Estimate and verify the maximun n that can be solved in 1 second



Analysis of Maximum n that can be solved in 1 second

First, let's look at the time complexity from the C++ code:
$$T(n) = 11 + 7n + [2(n-1)n+1, 2.5(n-1)n]$$
Let's expand the formula:

Lower bound: $2(n-1)n + 1 = 2n^2 - 2n + 1$
Upper bound: $2.5(n-1)n = 2.5n^2 - 2.5n$



Therefore, the complete range of instructions is:
$$T(n) = 11 + 7n + [2n^2 - 2n + 1, 2.5n^2 - 2.5n]$$

Assuming a modern computer can execute approximately $10^8$ operations per second, we can set up the inequality using the lower bound:

$$11 + 7n + (2n^2 - 2n + 1) \leq 10^8$$
Simplifying:
$$2n^2 + 5n + 12 \leq 10^8$$

Since the quadratic term dominates, we can approximate:
$$2n^2 \leq 10^8$$
$$n^2 \leq 5 \times 10^7$$
$$n \leq \sqrt{5 \times 10^7} \approx 7071$$

In [97]:
%%writefile test_generator.cpp
#include <iostream>
#include <fstream>
#include <random>
#include <string>

void generateTest(int n, const std::string& filename) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(1, 1000000000);  // Números entre 1 y 10^9

    std::ofstream file(filename);
    file << n << "\n";
    for (int i = 0; i < n; i++) {
        file << dis(gen);
        if (i < n-1) file << " ";
    }
    file << "\n";
    file.close();
}

int main() {
    // Generar tests con tamaños crecientes
    int sizes[] = {1000, 2000, 3000, 4000, 5000, 6000, 7000};

    for (int size : sizes) {
        std::string filename = "input_" + std::to_string(size) + ".txt";
        generateTest(size, filename);
        std::cout << "Generated " << filename << " with " << size << " numbers\n";
    }

    return 0;
}

Writing test_generator.cpp


In [98]:
!g++ test_generator.cpp -o test_generator
!./test_generator

# Compilar el programa principal
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long

# Ejecutar los tests y medir el tiempo
!time ./max_pairwise_product_long < input_1000.txt
!time ./max_pairwise_product_long < input_2000.txt
!time ./max_pairwise_product_long < input_3000.txt
!time ./max_pairwise_product_long < input_4000.txt
!time ./max_pairwise_product_long < input_5000.txt
!time ./max_pairwise_product_long < input_6000.txt
!time ./max_pairwise_product_long < input_7000.txt

Generated input_1000.txt with 1000 numbers
Generated input_2000.txt with 2000 numbers
Generated input_3000.txt with 3000 numbers
Generated input_4000.txt with 4000 numbers
Generated input_5000.txt with 5000 numbers
Generated input_6000.txt with 6000 numbers
Generated input_7000.txt with 7000 numbers
999686901002030670

real	0m0.007s
user	0m0.005s
sys	0m0.001s
998902235681819300

real	0m0.019s
user	0m0.014s
sys	0m0.003s
997447509693180975

real	0m0.034s
user	0m0.034s
sys	0m0.000s
999107194788782534

real	0m0.076s
user	0m0.070s
sys	0m0.001s
999552176314272944

real	0m0.085s
user	0m0.084s
sys	0m0.000s
999878096533696333

real	0m0.129s
user	0m0.123s
sys	0m0.002s
999667719326834250

real	0m0.173s
user	0m0.166s
sys	0m0.003s


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Timsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [99]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>

// Function to swap two elements in the vector
void swap(long long& a, long long& b) {
    long long temp = a;
    a = b;
    b = temp;
}

// Function to sort a vector using bubble sort implementation
void bubbleSort(std::vector<long long>& arr) {
    int n = arr.size();
    // Iterate through the array n-1 times
    for (int i = 0; i < n - 1; i++) {
        // Last i elements are already sorted
        for (int j = 0; j < n - i - 1; j++) {
            // Compare adjacent elements and swap if needed
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
            }
        }
    }
}

// Function to calculate the maximum pairwise product
long long MaxPairwiseProduct(const std::vector<long long>& numbers) {
    int n = numbers.size();

    // Create a copy of the original vector to avoid modifying it
    std::vector<long long> sorted_numbers(numbers);

    // Sort the vector using our custom bubble sort
    bubbleSort(sorted_numbers);

    // Calculate the product of the two largest numbers
    // (last two numbers in the sorted array)
    long long max_product = sorted_numbers[n - 1] * sorted_numbers[n - 2];

    return max_product;
}

int main() {
    // Declare variable to store the size of input
    int n;

    // Read the size of the array
    std::cin >> n;

    // Create a vector to store the numbers
    std::vector<long long> numbers(n);

    // Input the numbers from user
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    // Calculate and output the maximum pairwise product
    std::cout << MaxPairwiseProduct(numbers) << "\n";

    return 0;
}

Writing max_pairwise_product_sort.cpp


In [100]:
!g++ -o max_pairwise_product_sort max_pairwise_product_sort.cpp
!chmod +x max_pairwise_product_sort.cpp

Test with inputs 1 and 2



In [101]:
!./max_pairwise_product_sort < input1.txt
!./max_pairwise_product_sort < input2.txt

90
90


Test and time it with input 3

In [102]:
!time ./max_pairwise_product_sort < input3.txt

40000000000

real	0m0.004s
user	0m0.000s
sys	0m0.003s


20 stress tests

In [103]:
%%writefile generate_last_tests.cpp
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <ctime>
#include <string>

using namespace std;

// Function to generate a random long long number within a range
long long random_number(long long min, long long max) {
    long long range = max - min + 1;
    return min + (rand() % range);
}

// Function to create test file with specified parameters
void create_test_file(const string& filename, int n, long long min_val, long long max_val, bool sorted = false) {
    ofstream MyFile(filename);
    MyFile << n << "\n";

    // Create array to store numbers if we need to sort them
    long long* numbers = new long long[n];

    // Generate random numbers
    for (int i = 0; i < n; i++) {
        numbers[i] = random_number(min_val, max_val);
    }

    // If sorted flag is true, sort in descending order
    if (sorted) {
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (numbers[i] < numbers[j]) {
                    long long temp = numbers[i];
                    numbers[i] = numbers[j];
                    numbers[j] = temp;
                }
            }
        }
    }

    // Write numbers to file
    for (int i = 0; i < n; i++) {
        MyFile << numbers[i] << " ";
    }

    MyFile.close();
    delete[] numbers;
}

int main() {
    // Seed the random number generator
    srand(time(0));

    // input11.txt: 10000 random numbers between 1 and 100000
    create_test_file("input11.txt", 10000, 1, 100000);

    // input12.txt: 10000 random numbers between 1 and 100000
    create_test_file("input12.txt", 10000, 1, 100000);

    // input13.txt: 10000 random numbers between 1 and 100000
    create_test_file("input13.txt", 10000, 1, 100000);

    // input14.txt: 100000 random numbers between 1 and 1000000
    create_test_file("input14.txt", 100000, 1, 1000000);

    // input15.txt: 150000 random numbers between -1000000 and 1000000
    create_test_file("input15.txt", 150000, -1000000, 1000000);

    // input16.txt: 180000 random numbers, sorted in descending order
    create_test_file("input16.txt", 180000, 1, 2000000, true);

    // input17.txt: 190000 random numbers with some very large values
    create_test_file("input17.txt", 190000, -1000000000, 1000000000);

    // input18.txt: 195000 random numbers with many duplicates
    create_test_file("input18.txt", 195000, 1, 1000);

    // input19.txt: 198000 random numbers with mostly small values
    create_test_file("input19.txt", 198000, 0, 100);

    // input20.txt: 200000 random numbers with full range
    create_test_file("input20.txt", 200000, -2000000000, 2000000000);

    cout << "All test files have been generated successfully!" << endl;
    return 0;
}

Writing generate_last_tests.cpp


In [104]:
!g++ -o generate_last_tests generate_last_tests.cpp
!chmod +x generate_last_tests.cpp
!g++ -o generate_last_tests generate_last_tests.cpp
!./generate_last_tests

All test files have been generated successfully!


In [105]:
!./max_pairwise_product_sort < input1.txt
!./max_pairwise_product_sort < input2.txt
!./max_pairwise_product_sort < input3.txt
!./max_pairwise_product_sort < input4.txt
!./max_pairwise_product_sort < input5.txt
!./max_pairwise_product_sort < input6.txt
!./max_pairwise_product_sort < input7.txt
!./max_pairwise_product_sort < input8.txt
!./max_pairwise_product_sort < input9.txt
!./max_pairwise_product_sort < input10.txt
!./max_pairwise_product_sort < input11.txt
!./max_pairwise_product_sort < input12.txt
!./max_pairwise_product_sort < input13.txt
!./max_pairwise_product_sort < input14.txt
!./max_pairwise_product_sort < input15.txt
!./max_pairwise_product_sort < input16.txt
!./max_pairwise_product_sort < input18.txt
!./max_pairwise_product_sort < input19.txt
!./max_pairwise_product_sort < input20.txt

90
90
40000000000
913578400000
42000000000000
180000000000000
2250000000000000
13200000000000000
60000000000000000
1000000000000000000
9999500000
9996900108
9997400160
^C
^C
^C
^C
^C
^C


Count executed instructions in code and verify with inputs 1,2 and 3


In [106]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>

// Function to swap two elements in the vector
void swap(long long& a, long long& b) {
    long long temp = a;                    // 1
    a = b;                                 // 1
    b = temp;                              // 1
}                                          // Total: 3 instructions per swap

// Function to sort a vector using bubble sort implementation
void bubbleSort(std::vector<long long>& arr) {
    int n = arr.size();                    // 1
    for (int i = 0; i < n - 1; i++) {      // 2 + 2(n-1)
        for (int j = 0; j < n - i - 1; j++) { // (n-1) * [2 + 2(n-i-1)]
            if (arr[j] > arr[j + 1]) {      // (n-1)(n-i-1)
                swap(arr[j], arr[j + 1]);    // 3 * up to (n-1)(n-i-1)
            }
        }
    }
}                                          // Total: O(n²) with exact count:
                                          // 1 + 2(n-1) + Σᵢ₌₀ⁿ⁻²[2(n-i-1) + (n-i-1)(1 + 3)]

// Function to calculate the maximum pairwise product
long long MaxPairwiseProduct(const std::vector<long long>& numbers) {
    int n = numbers.size();                // 1

    std::vector<long long> sorted_numbers(numbers); // n (copy construction)

    bubbleSort(sorted_numbers);            // See bubbleSort analysis above

    long long max_product = sorted_numbers[n - 1] * sorted_numbers[n - 2]; // 3

    return max_product;                    // 1
}                                         // Total: 5 + n + bubbleSort complexity

int main() {
    int n;                                // 1
    std::cin >> n;                        // 1

    std::vector<long long> numbers(n);    // n

    for (int i = 0; i < n; ++i) {        // 2 + 2n
        std::cin >> numbers[i];           // n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n"; // 1 + MaxPairwiseProduct complexity

    return 0;                            // 1
}                                        // Total: 6 + 4n + MaxPairwiseProduct complexity

Overwriting max_pairwise_product_sort.cpp


Let's analyze the number of instructions executed for each input case. For our algorithm using bubble sort, the general formula for executed instructions is:
$$T(n) = 12 + 7n + n(n-1)(4) = 12 + 7n + 4n^2 - 4n = 4n^2 + 3n + 12$$
This formula comes from:

12 base instructions (initializations, returns, etc.)
7n from linear operations (vector copies, input loops)
4n² - 4n from bubble sort (comparisons and swaps)

for input1.txt

In this case (best case, already sorted list):

Comparisons: 45 (n(n-1)/2)
Swaps: 0 (already sorted)
Total instructions: 12 + 7(10) + 10(9)(1) = 12 + 70 + 90 = 172

for input2.txt

Comparisons: 45 (n(n-1)/2)
Swaps: 45 (each comparison results in a swap)
Total instructions: 12 + 7(10) + 10(9)(4) = 12 + 70 + 360 = 442

for input3.txt

Comparisons: 15 (n(n-1)/2)
Swaps: ~8 (estimated based on unsorted elements)
Total instructions: 12 + 7(6) + 6(5)(2.5) = 12 + 42 + 75 = 129

Calculate estimated time by instruction and number of instructions by second

In [107]:
n = 50000
numbers = " ".join(map(str, range(1, 50001)))
with open('input21.txt', 'w') as file:
    file.write(f"{n}\n{numbers}\n")

In [108]:
!time ./max_pairwise_product_sort < input21.txt

2499950000

real	0m8.871s
user	0m8.707s
sys	0m0.012s


Time and Instruction Analysis for n = 50000
Total Instructions Calculation
For n = 50000 (sorted case - best case scenario):
T(n) = 12 + 7n + n(n-1)(1)
= 12 + 7(50000) + 50000(49999)(1)
= 12 + 350000 + 2499950000
= 2500300012 instructions
Time Analysis
Observed execution time: 8.103 seconds
Time per Instruction

Time per instruction = Total time / Total instructions
= 8.103 seconds / 2500300012 instructions
= 3.24 × 10⁻⁹ seconds per instruction (or 3.24 nanoseconds per instruction)

Instructions per Second

Instructions per second = Total instructions / Total time
= 2500300012 / 8.103
≈ 308,564,976 instructions per second (or about 308.6 million instructions per second)

### Estimate and verify the maximun n that can be solved in 1 second

Estimating Maximum n for 1 Second Execution
Step 1: Calculate Instructions per Second
From our previous analysis with n = 50000:

Time taken: 8.103 seconds
Total instructions: 2500300012
Instructions per second ≈ 308,564,976

Step 2: Solve the Equation
For our algorithm:

T(n) = 4n² + 3n + 12 instructions (worst case)
We want T(n) instructions to be executed in 1 second
Therefore: 4n² + 3n + 12 = 308,564,976

Solving this quadratic equation:

4n² + 3n + (12 - 308,564,976) = 0
4n² + 3n - 308,564,964 = 0

Using quadratic formula:

n = [-3 ± √(9 + 4(4)(308,564,964))] / 8
n = [-3 ± √(4,937,039,431)] / 8
n = [-3 ± 70,264] / 8
n ≈ 8,782.625

In [109]:
n = 8782
numbers = " ".join(map(str, range(1, 8783)))
with open('verify_n.txt', 'w') as file:
    file.write(f"{n}\n{numbers}\n")

In [110]:
!time ./max_pairwise_product_sort < verify_n.txt

77114742

real	0m0.247s
user	0m0.241s
sys	0m0.002s


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [111]:
%%writefile max_pairwise_product_python.py
def max_pairwise_product(numbers):
    # Get length of array
    # Cost: 1 instruction
    n = len(numbers)

    # Initialize first maximum index
    # Cost: 1 instruction
    max1_index = 0

    # Find index of maximum number
    # Cost: 2n instructions (n comparisons + n iterations)
    for i in range(n):
        if numbers[i] > numbers[max1_index]:
            max1_index = i

    # Initialize second maximum index to first number if max1 is second number,
    # otherwise initialize to second number
    # Cost: 2 instructions
    max2_index = 0 if max1_index == 1 else 1

    # Find second largest number, skip max1_index
    # Cost: 2n instructions (n comparisons + n iterations)
    for i in range(n):
        if i != max1_index and numbers[i] > numbers[max2_index]:
            max2_index = i

    # Return product of two largest numbers
    # Cost: 1 instruction
    return numbers[max1_index] * numbers[max2_index]

# Read size of input
# Cost: 1 instruction
n = int(input())

# Read and convert input string to list of integers
# Cost: n instructions
input_numbers = list(map(int, input().split()))

# Calculate maximum pairwise product
# Cost: 1 instruction (call) + function cost
result = max_pairwise_product(input_numbers)

# Print result
# Cost: 1 instruction
print(result)

# Total instruction count analysis:
# Base operations: 7 instructions (initializations, return, input/output)
# Loop operations: 5n instructions
#   - 2n for first loop (finding max1)
#   - 2n for second loop (finding max2)
#   - n for input processing
# Total: T(n) = 5n + 7 instructions

Writing max_pairwise_product_python.py


Test with inputs 1 and 2



In [112]:
!chmod +x max_pairwise_product_python.py
!ls -l max_pairwise_product_python.py

-rwxr-xr-x 1 root root 1582 Dec  8 23:05 max_pairwise_product_python.py


In [113]:
!python ./max_pairwise_product_python.py < input1.txt
!python ./max_pairwise_product_python.py < input2.txt

90
90


Test and time it with input 3

In [114]:
!python ./max_pairwise_product_python.py < input3.txt

40000000000


20 stress tests

In [115]:
!python ./max_pairwise_product_python.py < input1.txt
!python ./max_pairwise_product_python.py < input2.txt
!python ./max_pairwise_product_python.py < input3.txt
!python ./max_pairwise_product_python.py < input4.txt
!python ./max_pairwise_product_python.py < input5.txt
!python ./max_pairwise_product_python.py < input6.txt
!python ./max_pairwise_product_python.py < input7.txt
!python ./max_pairwise_product_python.py < input8.txt
!python ./max_pairwise_product_python.py < input9.txt
!python ./max_pairwise_product_python.py < input10.txt
!python ./max_pairwise_product_python.py < input11.txt
!python ./max_pairwise_product_python.py < input12.txt
!python ./max_pairwise_product_python.py < input13.txt
!python ./max_pairwise_product_python.py < input14.txt
!python ./max_pairwise_product_python.py < input15.txt
!python ./max_pairwise_product_python.py < input16.txt
!python ./max_pairwise_product_python.py < input17.txt
!python ./max_pairwise_product_python.py < input18.txt
!python ./max_pairwise_product_python.py < input19.txt
!python ./max_pairwise_product_python.py < input20.txt

90
90
40000000000
913578400000
42000000000000
180000000000000
2250000000000000
13200000000000000
60000000000000000
1000000000000000000
9999500000
9996900108
9997400160
999999000000
999932001152
3999992000003
999923120037658488
1000000
10000
21746249744779648


Count executed instructions in code and verify with inputs 1,2 and 3


for input1.txt

In this case:

n = 10
Instructions = 5(10) + 7
Instructions = 50 + 7 = 57 instructions

Breakdown:

Finding max1_index (max value = 10 at index 9):

10 iterations + 10 comparisons = 20 instructions


Finding max2_index (second max = 9 at index 8):

10 iterations + 10 comparisons = 20 instructions


Input processing: 10 instructions
Base operations: 7 instructions


Total: 57 instructions

for input2.txt

In this case:

n = 10
Instructions = 5(10) + 7
Instructions = 57 instructions

Breakdown:

Finding max1_index (max value = 10 at index 0):

10 iterations + 10 comparisons = 20 instructions


Finding max2_index (second max = 9 at index 1):

10 iterations + 10 comparisons = 20 instructions


Input processing: 10 instructions
Base operations: 7 instructions


Total: 57 instructions

 for input3.txt

 In this case:

n = 6
Instructions = 5(6) + 7
Instructions = 30 + 7 = 37 instructions

Breakdown:

Finding max1_index (max value = 200000 at index 2):

6 iterations + 6 comparisons = 12 instructions


Finding max2_index (second max = 200000 at index 4):

6 iterations + 6 comparisons = 12 instructions


Input processing: 6 instructions
Base operations: 7 instructions


Total: 37 instructions

Calculate estimated time by instruction and number of instructions by second

In [116]:
!time python ./max_pairwise_product_python.py < input21.txt

2499950000

real	0m0.063s
user	0m0.047s
sys	0m0.014s



Time and Instruction Analysis for Python Implementation (n = 50000)

For n = 50000:
T(n) = 5n + 7
T(50000) = 5(50000) + 7
= 250000 + 7
= 250007 instructions

Measured Execution Time

Total time (real): 0.060 seconds
CPU user time: 0.044 seconds
CPU system time: 0.015 seconds

Instruction Count
For n = 50000:

T(n) = 5n + 7
T(50000) = 5(50000) + 7
= 250,007 instructions total

Time per Instruction Calculation
Using total execution time:

Time per instruction = Total time / Total instructions
= 0.060 seconds / 250,007 instructions
= 2.4 × 10⁻⁷ seconds per instruction
= 240 nanoseconds per instruction

Instructions per Second Calculation

Instructions per second = Total instructions / Total time
= 250,007 / 0.060
≈ 4,166,783 instructions per second
≈ 4.17 million instructions per second

### Estimate and verify the maximun n that can be solved in 1 second

From our last test with n = 50000:

Time taken: 0.060 seconds
Instructions executed: 5(50000) + 7 = 250,007
Instructions per second ≈ 4,166,783

Step 2: Calculate Maximum n
Using our formula T(n) = 5n + 7 and our measured execution speed:

For 1 second, we can execute ≈ 4,166,783 instructions
5n + 7 = 4,166,783
5n = 4,166,776
n ≈ 833,355

Step 3: Add Safety Margin
Considering:

System overhead
Input/output operations
Memory constraints
Python's garbage collection

Let's use n = 800,000 for verification

In [117]:
# Create test file for n = 800000
n = 800000
numbers = " ".join(map(str, range(1, n+1)))
with open('verify_max_n.txt', 'w') as file:
    file.write(f"{n}\n{numbers}\n")

In [118]:
!time python ./max_pairwise_product_python.py < verify_max_n.txt

639999200000

real	0m0.498s
user	0m0.394s
sys	0m0.098s


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [119]:
%%writefile super_efficient_python.py
n = int(input())                    # 1 instruction
max1, max2 = float('-inf'), float('-inf')  # 2 instructions
numbers = map(int, input().split()) # 1 instruction (iterator creation)

for num in numbers:                 # n iterations
    if num > max1:                  # n comparisons
        max2 = max1                 # up to n assignments
        max1 = num                  # up to n assignments
    elif num > max2:                # up to n comparisons
        max2 = num                  # up to n assignments

print(max1 * max2)                  # 1 instruction

# Total: 5 + 6n instructions

Writing super_efficient_python.py


In [120]:
!chmod +x super_efficient_python.py
!ls -l super_efficient_python.py

-rwxr-xr-x 1 root root 603 Dec  8 23:05 super_efficient_python.py


Test with inputs 1 and 2




In [121]:
!python ./super_efficient_python.py < input1.txt
!python ./super_efficient_python.py < input2.txt

90
90


Test and time it with input 3

In [122]:
!python ./super_efficient_python.py < input3.txt

40000000000


20 stress tests

In [123]:
!python ./super_efficient_python.py < input1.txt
!python ./super_efficient_python.py < input2.txt
!python ./super_efficient_python.py < input3.txt
!python ./super_efficient_python.py < input4.txt
!python ./super_efficient_python.py < input5.txt
!python ./super_efficient_python.py < input6.txt
!python ./super_efficient_python.py < input7.txt
!python ./super_efficient_python.py < input8.txt
!python ./super_efficient_python.py < input9.txt
!python ./super_efficient_python.py < input10.txt
!python ./super_efficient_python.py < input11.txt
!python ./super_efficient_python.py < input12.txt
!python ./super_efficient_python.py < input13.txt
!python ./super_efficient_python.py < input14.txt
!python ./super_efficient_python.py < input15.txt
!python ./super_efficient_python.py < input16.txt
!python ./super_efficient_python.py < input17.txt
!python ./super_efficient_python.py < input18.txt
!python ./super_efficient_python.py < input19.txt
!python ./super_efficient_python.py < input20.txt

90
90
40000000000
913578400000
42000000000000
180000000000000
2250000000000000
13200000000000000
60000000000000000
1000000000000000000
9999500000
9996900108
9997400160
999999000000
999932001152
3999992000003
999923120037658488
1000000
10000
21746249744779648


Count executed instructions in code and verify with inputs 1,2 and 3

for input1.txt

Instructions executed:

Initial setup: 4 instructions
Loop analysis (n = 10):

Iterations: 10
All numbers cause max1 update: 10 * 4 instructions (comparison + 2 assignments + iteration)
No max2-only updates


Final print: 1 instruction

Total = 4 + (10 * 4) + 1 = 45 instructions

for input2.txt

Instructions executed:

Initial setup: 4 instructions
Loop analysis (n = 10):

First number (10): 4 instructions
Remaining 9 numbers: 9 * 3 instructions (comparison + 1 assignment + iteration)


Final print: 1 instruction

Total = 4 + 4 + (9 * 3) + 1 = 32 instructions

for input3.txt

Instructions executed:

Initial setup: 4 instructions
Loop analysis (n = 6):

Numbers updating max1 (200000): 4 instructions twice
Numbers updating max2: 3 instructions once
Numbers not updating either: 3 instructions thrice


Final print: 1 instruction

Total = 4 + (2 * 4) + (1 * 3) + (3 * 3) + 1 = 22 instructions

Calculate estimated time by instruction and number of instructions by second

In [124]:
!time python ./super_efficient_python.py < input21.txt

2499950000

real	0m0.125s
user	0m0.067s
sys	0m0.018s


Time and Instruction Analysis for n = 50000
Execution Time Measurements

Real time: 0.054 seconds
User CPU time: 0.038 seconds
System CPU time: 0.016 seconds

Calculate Total Instructions
For n = 50000:

T(n) = 5 + 4n instructions
T(50000) = 5 + 4(50000)
= 5 + 200000
= 200,005 total instructions

Time per Instruction
Using total execution time:

Time per instruction = Total time / Total instructions
= 0.054 seconds / 200,005 instructions
= 2.7 × 10⁻⁷ seconds per instruction
= 270 nanoseconds per instruction

Instructions per Second

Instructions per second = Total instructions / Total time
= 200,005 / 0.054
≈ 3,703,796 instructions per second
≈ 3.7 million instructions per second

### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [125]:
%%writefile super_efficient.py
# Read the size of input
n = int(input())                                    # 1 instruction: integer conversion

# Initialize the two largest numbers as minimum possible values
max1 = 0                                           # 1 instruction: assignment
max2 = 0                                           # 1 instruction: assignment

# Process each number from input without storing the list
for num in map(int, input().split()):             # n iterations: split, convert and iterate
    if num > max1:                                # n comparisons
        max2 = max1                               # up to n assignments
        max1 = num                                # up to n assignments
    elif num > max2:                              # up to n comparisons
        max2 = num                                # up to n assignments

# Print the result
print(max1 * max2)                                # 1 instruction: multiplication and print

# Total instruction count breakdown:
# - Base instructions (initialization + final print): 4
# - Per number processing: 4 instructions maximum per number
# - Total worst case: T(n) = 4 + 4n instructions
# - Total best case: T(n) = 4 + 2n instructions (when numbers are in descending order)
# Therefore: T(n) ≤ 4 + 4n = O(n)


Overwriting super_efficient.py


Test with inputs 1 and 2




In [126]:
!python ./super_efficient.py < input1.txt
!python ./super_efficient.py < input2.txt

90
90


Test and time it with input 3

In [128]:
!python ./super_efficient.py < input3.txt

40000000000


20 stress tests

In [130]:
!python ./super_efficient.py < input1.txt
!python ./super_efficient.py < input2.txt
!python ./super_efficient.py < input3.txt
!python ./super_efficient.py < input4.txt
!python ./super_efficient.py < input5.txt
!python ./super_efficient.py < input6.txt
!python ./super_efficient.py < input7.txt
!python ./super_efficient.py < input8.txt
!python ./super_efficient.py < input9.txt
!python ./super_efficient.py < input10.txt
!python ./super_efficient.py < input11.txt
!python ./super_efficient.py < input12.txt
!python ./super_efficient.py < input13.txt
!python ./super_efficient.py < input14.txt
!python ./super_efficient.py < input15.txt
!python ./super_efficient.py < input16.txt
!python ./super_efficient.py < input17.txt
!python ./super_efficient.py < input18.txt
!python ./super_efficient.py < input19.txt
!python ./super_efficient.py < input20.txt

90
90
40000000000
913578400000
42000000000000
180000000000000
2250000000000000
13200000000000000
60000000000000000
1000000000000000000
9999500000
9996900108
9997400160
999999000000
999932001152
3999992000003
999923120037658488
1000000
10000
21746249744779648


Count executed instructions in code and verify with inputs 1,2 and 3

For Input 1

Instruction count:

Initial setup:

n = int(input()): 1 instruction
max1 = 0: 1 instruction
max2 = 0: 1 instruction


Loop execution (ascending numbers):

First number (1):

comparison with max1: 1
max2 = max1: 1
max1 = num: 1
iteration: 1


Each subsequent number (2-10):

Same pattern as all numbers are larger than max1
4 instructions × 9 numbers = 36 instructions


Total = 3 + 4 + 36 + 1 (final print) = 44 instructions

For input 2

Instruction count:

Initial setup: 3 instructions
Loop execution (descending numbers):

First number (10):

comparison with max1: 1
max2 = max1: 1
max1 = num: 1
iteration: 1


Each subsequent number:

comparison with max1: 1
comparison with max2: 1
iteration: 1
max2 update when needed: 1
3 instructions × 9 numbers = 27 instructions

Total = 3 + 4 + 27 + 1 (final print) = 35 instructions

For input 3

Instruction count:

Initial setup: 3 instructions
Loop execution:

First number (1): 4 instructions
Second number (3): 4 instructions
Third number (200000): 4 instructions
Fourth number (9): 3 instructions (comparison + comparison with max2)
Fifth number (200000): 4 instructions
Sixth number (23): 3 instructions

Total = 3 + (4×3 + 3×2) + 1 (final print) = 25 instructions

Calculate estimated time by instruction and number of instructions by second

In [133]:
!time python ./super_efficient.py < input21.txt

2499950000

real	0m0.057s
user	0m0.040s
sys	0m0.016s


Time and Instruction Analysis for n = 50000
Measured Execution Time

    Real time: 0.074 seconds
    User CPU time: 0.050 seconds
    System time: 0.021 seconds

Calculate Total Instructions

For n = 50000:

    T(n) = 4 + 4n instructions (worst case)
    T(50000) = 4 + 4(50000)
    = 4 + 200000
    = 200,004 total instructions

Time per Instruction

    Time per instruction = Total time / Total instructions
    = 0.074 seconds / 200,004 instructions
    = 3.7 × 10⁻⁷ seconds per instruction
    = 370 nanoseconds per instruction


### Estimate and verify the maximun n that can be solved in 1 second

Step 1: Using Previous Measurement
From our last test (n = 50000):

Time taken: 0.074 seconds
Instructions executed: 4 + 4(50000) = 200,004
Instructions per second ≈ 2,702,756

Step 2: Calculate Maximum n
Using our formula T(n) = 4 + 4n and measured execution speed:

For 1 second, we can execute ≈ 2,702,756 instructions
4 + 4n = 2,702,756
4n = 2,702,752
n ≈ 675,688

Step 3: Add Safety Margin
Considering:

System overhead (28.4% in previous test)
Input/output operations
Python interpreter overhead
Memory constraints

Let's use n = 650,000 for verification

In [132]:
# Create test file for n = 650000
n = 650000
numbers = " ".join(map(str, range(1, n+1)))
with open('verify_max_n.txt', 'w') as file:
    file.write(f"{n}\n{numbers}\n")

In [134]:
!time python ./super_efficient.py < verify_max_n.txt

422499350000

real	0m0.285s
user	0m0.213s
sys	0m0.070s


## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".